include/deal.II/lac/petsc_snes.h

```diff
@@ -0,0 +1,477 @@
+     *
+     * This function returns the number of iterations.
+     * The vector @p x must contain the initial guess.
+     * Upon returning, the @p x vector contains the solution.
+     */
+    unsigned int
+    solve(VectorType &x);
+
+    /**
+     * Solve the nonlinear system of equations $F(x) = 0$.
+     *
+     * This function returns the number of iterations.
+     * The vector @p x must contain the initial guess.
+     * Upon returning, the @p x vector contains the solution.
+     *
+     * Here we also set the matrix to handle the Jacobian.
+     */
+    unsigned int
+    solve(VectorType &x, PMatrixType &P);
+
+    /**
+     * Solve the nonlinear system of equations $F(x) = 0$.
+     *
+     * This function returns the number of iterations.
+     * The vector @p x must contain the initial guess.
+     * Upon returning, the @p x vector contains the solution.
+     *
+     * Here we also set the matrices to handle Jacobians.
+     */
+    unsigned int
+    solve(VectorType &x, AMatrixType &A, PMatrixType &P);
+
+    /**
+     * Solve the affine nonlinear system of equations $F(x) = b$.
+     *
+     * This function returns the number of iterations.
+     * The vector @p x must contain the initial guess.
+     * Upon returning, the @p x vector contains the solution.
+     *
+     * Upon returning, the @p x vector contains the solution.
+     */
+    unsigned int
+    solve(VectorType &x, const VectorType &b);
+
+    /**
+     * Solve the affine nonlinear system of equations $F(x) = b$.
+     *
+     * This function returns the number of iterations.
+     * The vector @p x must contain the initial guess.
+     * Upon returning, the @p x vector contains the solution.
+     *
+     * Here we also set the matrix to handle the Jacobian.
+     */
+    unsigned int
+    solve(VectorType &x, const VectorType &b, PMatrixType &P);
+
+    /**
+     * Solve the affine nonlinear system of equations $F(x) = b$.
+     *
+     * This function returns the number of iterations.
+     * The vector @p x must contain the initial guess.
+     * Upon returning, the @p x vector contains the solution.
+     *
+     * Here we also set the matrices to handle Jacobians.
+     */
+    unsigned int
+    solve(VectorType &x, const VectorType &b, AMatrixType &A, PMatrixType &P);
+
+    /**
+     * Callback for the computation of the nonlinear residual $F(x)$.
+     */
+    std::function<int(const VectorType &x, VectorType &res)> residual;
+
+    /**
+     * Callback for the computation of the Jacobian
+     * $\dfrac{\partial F}{\partial x}$
+     */
+    std::function<int(const VectorType &x, AMatrixType &A, PMatrixType &P)>
+      jacobian;
+
+    /**
+     * Callback for monitoring the solution process.
+     *
+     * This function is called by NonlinearSolver at the beginning
+     * of each time step.
+     */
+    std::function<
+      int(const VectorType &x, const unsigned int step_number, real_type fval)>
```

Comment:
What is `fval`?

Reply:
current value of the residual. I will fix the documentation. Thanks